api/tools/src/main/java/org/pmiops/workbench/tools/BulkCloneWorkspacesTool.java

```diff
@@ -0,0 +1,329 @@
+      return false;
+    }
+  }
+
+  @Bean
+  public CommandLineRunner run(
+      WorkspaceDao workspaceDao,
+      WorkspacesController workspacesController,
+      WorkspaceService workspaceService,
+      FireCloudService fireCloudService,
+      UserDao userDao,
+      BillingProjectBufferService billingProjectBufferService,
+      DirectoryService directoryService,
+      @Qualifier("workspaceAclsApi") WorkspacesApi saWorkspaceApi) {
+    return (args) -> {
+      padding();
+      initializeApis();
+      System.out.println("Apis initialized");
+
+      int numToProcess = Integer.parseInt(args[0]);
+
+      boolean dryRun = false;
+      if (args.length > 1 && args[1].equals("true")) {
+        dryRun = true;
+      }
+
+      List<WorkspaceResponse> processed = new ArrayList<>();
+      List<Pair<WorkspaceResponse, String>> failedWorkspaces = new ArrayList<>();
+
+      for (WorkspaceResponse workspaceResponse : saWorkspaceApi.listWorkspaces()) {
+        Workspace dbWorkspace =
+            workspaceDao.findByFirecloudUuid(workspaceResponse.getWorkspace().getWorkspaceId());
+
+        // Workspace that exists in FC but not in AoU
+        if (dbWorkspace == null) {
+          continue;
+        }
+
+        // Skip over inactive workspaces and workspaces that are already on the
+        if (dbWorkspace.getWorkspaceActiveStatusEnum().equals(WorkspaceActiveStatus.DELETED)) {
+          continue;
+        }
+
+        // Only process workspaces that need migration
+        if (!dbWorkspace.getBillingMigrationStatusEnum().equals(BillingMigrationStatus.OLD)) {
+          continue;
+        }
+
+        if (workspaceResponse.getAccessLevel().equals("NO ACCESS")) {
+          System.out.println(
+              "Found a workspace that the SA account cannot access : "
+                  + shorthand(workspaceResponse.getWorkspace()));
+          failedWorkspaces.add(Pair.of(workspaceResponse, "NO ACCESS"));
+          continue;
+        }
+
+        final double bufferThresholdPercentage = 25;
+        final int sleepIntervalSeconds = 30;
+        while (billingProjectBufferService.availableProportion()
+            < bufferThresholdPercentage / 100) {
+          System.out.println(
+              "Less than "
+                  + bufferThresholdPercentage
+                  + "% of the buffer is available ("
+                  + billingProjectBufferService.availableProportion() * 100
+                  + "%)... Sleeping for "
+                  + sleepIntervalSeconds
+                  + " seconds");
+          Thread.sleep(sleepIntervalSeconds * 1000);
+        }
+
+        providedUser = userDao.findUserByEmail(workspaceResponse.getWorkspace().getCreatedBy());
+        if (providedUser == null || providedUser.getDisabled() || !checkUserExistsInGoogle(directoryService, providedUser.getEmail())) {
+          continue;
+        }
+
+        impersonateUser(fireCloudService.getApiClientWithImpersonation(providedUser.getEmail()));
```

Comment:
Note: One follow-up, if we want to establish this as a pattern in our tools we'll need to figure out a way to do this without mutating shared state like this. The next API call made by any dependency will now utilize this impersonated user cred, which is scary and could easily be unintentional.